### PR TITLE
gitrepo: don't fallback to origin when resolving version

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/7678

There are also some corner cases defined and tested:

- Tested there is no fallback to resolve origin branches in ResolveVersion.
- Test resolving for the detached head.
- Test handling unknown reference.
- Renamed IsNotFound to IsReferenceNotFound
- referenceNotFoundError may be returned by ResolveVersion

### CODE REVIEW

File: pkg/gitrepo/repo_test.go [hunk 1]
```diff
@@ -251,6 +279,21 @@ func Test_Repo_ResolveVersion(t *testing.T) {
...
+				if err != nil {
+					t.Fatalf("err = %v, want %v", microerror.Stack(err), nil)
+				}
+			}
+
+			h, err := repo.HeadSHA(context.Background())
+			if err != nil {
+				t.Fatalf("err = %v, want %v", microerror.Stack(err), nil)
+			}
+			t.Logf("head=%s", h)
```
Comment (xh3b4sd) [inline pkg/gitrepo/repo_test.go]: What is with that?

Comment (kopiczko) [reply pkg/gitrepo/repo_test.go]: My debugging leftover. I'm removing it. Thanks.
